### DISCUSSION TITLE
Metadata for Typst documents

### DISCUSSION THREAD

**Discussion by nandac** [jgm/pandoc#9937] [Q&A]
Dear Folks,

I am trying to add support for Typst documents for a plugin used by the Pelican Static Site Generator.

The plugin only supports Markdown variants and I would like to add support to convert Typst to HTML5.

The issue I am having is how to add metadata to a Typst document without using a YAML block. How does Typst handle metadata? How does the whole metadata thing work with Pandoc?

Any insights on how this works would be greatly appreciated.

Many thanks

**Comment by iandol** [comment]
Typst doesn't use metadata per se, it uses Typst templates and you complete the function by passing Typst arrays (see https://github.com/jgm/pandoc-templates/blob/master/default.typst#L23 for how the title/author etc. get into the typst output). If you need to extract these values, you would probably need a preparser of some kind; you can then use defaults and/or metadata passed to Pandoc for the conversion to HTML?

**Comment by nandac** [comment]
Thanks for your reply. I was thinking I would have to add these values in a
separate file and pass them in using metadata-file.

I like the pre-parsing method a lot more as it reduces the need for a
separate file.

**Reply by bpj** [reply]
Which language are you writing the plugin in? It is probably pretty easy to match a YAML block at the top of the file, write it to a temp file and pass that to pandoc with `--metadata-file`. How to match it depends on the available regex/pattern flavor though. In principle match `---` on its own line, then non-greedy matching for whatever up to `...` or another `---` on its own line.

Another option may be to place a table structure native to the format at the top of the file (perhaps wrapped in something which becomes a Div with a class "meta" in Pandoc's AST) and  process it with a Pandoc filter

**Reply by nandac** [reply]
My [plugin](https://github.com/pelican-plugins/pandoc-reader) is written in Python and I currently parse the YAML block at the top of the file exactly as you have mentioned.

Having a YAML block at the top of a Typst file is not a great option as it is not standard Typst.

I am interested in knowing how the Typst reader handles metadata.

Are users required to call the default Typst template to send the title, author and date if they want to use Pandoc to convert to HTML?

**Reply by iandol** [reply]
> I am interested in knowing how the Typst reader handles metadata.

It doesn't as there is no clear concept of metadata in Typst. Look at their advanced tutorial (https://typst.app/docs/tutorial/advanced-styling/) and you'll see the title/author data is part of the document content itself:

```
#grid(
  columns: (1fr, 1fr),
  align(center)[
    Therese Tungsten \
    Artos Institute \
    #link("mailto:tung@artos.edu")
  ],
  align(center)[
    Dr. John Doe \
    Artos Institute \
    #link("mailto:doe@artos.edu")
  ]
)
```

There are also templates where you pass arrays for more general use, but the template itself defines the fields, there is no universal values. For example:

```typst
#show: doc => conf(
  title: [Towards Improved Modelling],
  authors: (
    (
      name: "Theresa Tungsten",
      affiliation: "Artos Institute",
      email: "tung@artos.edu",
    ),
    (
      name: "Eugene Deklan",
      affiliation: "Honduras State",
      email: "e.deklan@hstate.hn",
    ),
  ),
  abstract: lorem(80),
  doc,
)
```
https://typst.app/docs/tutorial/making-a-template/#named-arguments

I think you need to define what kinds of typst documents you want to accept, before you will now what steps you need to improve the conversion.

> Are users required to call the default Typst template to send the title, author and date if they want to use Pandoc to convert to HTML?

You would not be using the default (or even a custom) Typst template if you are converting FROM typst TO html. For the conversion to html you could either leave the document contents in place, i.e. in the in-doc author example above you could hope the grid itself converted to html ok without needing to extract the names, then make them metadata. Pandoc's html template does usually work with metadata, and you may need to tweak the html template to accomadate any typst oddities (with custom CSS)
